src/controller/event/handler/webhook/artifact/retention.go

```diff
@@ -0,0 +1,148 @@
+	}
+	if len(trEvent.Deleted) == 0 {
+		log.Debugf("empty delete info of retention event")
+		return nil
+	}
+
+	payload, dryRun, project, err := constructRetentionPayload(trEvent)
+	if err != nil {
+		return err
+	}
+	// if dry run, do not trigger webhook
+	if dryRun {
+		log.Debugf("retention task %v is dry run", trEvent.TaskID)
+		return nil
+	}
+
+	policies, err := notification.PolicyMgr.GetRelatedPolices(project, trEvent.EventType)
+	if err != nil {
+		log.Errorf("failed to find policy for %s event: %v", trEvent.EventType, err)
+		return err
+	}
+	if len(policies) == 0 {
+		log.Debugf("cannot find policy for %s event: %v", trEvent.EventType, trEvent)
+		return nil
+	}
+	err = util.SendHookWithPolicies(policies, payload, trEvent.EventType)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// IsStateful ...
+func (r *RetentionHandler) IsStateful() bool {
+	return false
+}
+
+func constructRetentionPayload(event *event.RetentionEvent) (*model.Payload, bool, int64, error) {
+	task, err := api.RetentionController.GetRetentionExecTask(event.TaskID)
+	if err != nil {
+		log.Errorf("failed to get retention task %d: error: %v", event.TaskID, err)
+		return nil, false, 0, err
+	}
+	if task == nil {
+		return nil, false, 0, fmt.Errorf("task %d not found with retention event", event.TaskID)
+	}
+
+	execution, err := api.RetentionController.GetRetentionExec(task.ExecutionID)
+	if err != nil {
+		log.Errorf("failed to get retention execution %d: error: %v", task.ExecutionID, err)
+		return nil, false, 0, err
+	}
+	if execution == nil {
+		return nil, false, 0, fmt.Errorf("execution %d not found with retention event", task.ExecutionID)
+	}
+
+	if execution.DryRun {
+		return nil, true, 0, nil
+	}
+
+	md, err := api.RetentionController.GetRetention(execution.PolicyID)
+	if err != nil {
+		log.Errorf("failed to get tag retention policy %d: error: %v", execution.PolicyID, err)
+		return nil, false, 0, err
+	}
+	if md == nil {
+		return nil, false, 0, fmt.Errorf("policy %d not found with tag retention event", execution.PolicyID)
+	}
+
+	extURL, err := config.ExtURL()
+	if err != nil {
+		log.Errorf("Error while reading external endpoint URL: %v", err)
+	}
+	hostname := strings.Split(extURL, ":")[0]
+
+	payload := &model.Payload{
+		Type:     event.EventType,
+		OccurAt:  event.OccurAt.Unix(),
+		Operator: execution.Trigger,
+		EventData: &model.EventData{
+			Retention: &ctlModel.Retention{
+				Total:             task.Total,
+				Retained:          task.Retained,
+				HarborHostname:    hostname,
+				ProjectName:       event.Deleted[0].Target.Namespace,
+				RetentionPolicyID: execution.PolicyID,
+				Status:            event.Status,
+				RetentionRules:    []*ctlModel.RetentionRule{},
+			},
+		},
+	}
+
+	for _, v := range event.Deleted {
+		target := v.Target
+		deletedArtifact := &ctlModel.ArtifactInfo{
+			Type:       target.Kind,
+			Status:     event.Status,
+			NameAndTag: target.Repository + ":" + target.Tags[0],
```

Comment:
be careful the `panic`(runtime error: index out of range) here.

Reply:
> be careful the `panic`(runtime error: index out of range) here.

fixed